legate/core/runtime.py

```diff
@@ -1334,6 +1374,20 @@ def get_barriers(self, count: int) -> tuple[Future, Future]:
             Future.from_cdata(self.legion_runtime, wait_barrier),
         )
 
+    def record_pending_exception(
+        self,
+        exn_types: list[type],
+        future: Future,
+        tb: Optional[TracebackType] = None,
+    ) -> None:
+        exn = PendingException(exn_types, future, tb)
+        self._pending_exceptions.append(exn)
+
+    def raise_exceptions(self) -> None:
+        for pending in self._pending_exceptions:
+            pending.raise_exception()
+        self._pending_exceptions = []
```

Comment:
Move this before the for-loop:

```
     def raise_exceptions(self) -> None:
-        for pending in self._pending_exceptions:
-            pending.raise_exception()
+        pending_exceptions = self._pending_exceptions
         self._pending_exceptions = []
+        for pending in pending_exceptions:
+            pending.raise_exception()
```

With the current implementation, if any of the `pending` operations throws, then the `_pending_exceptions` list doesn't get cleaned, so even if the exception is caught and execution proceeds, the exception will be thrown again at shutdown time, when legate calls `raise_exceptions` again.

Reply:
A good catch. b860e8d9479b5c5fb4eb0cd3535bed237d4e8c12 I wonder whether we should add an API call for clients to manually block on exceptions. I feel client would need a programmable way to control how exceptions would be handled. that would also make exception tests for client libraries easier to write. for example, cuNumeric could (re-)export the blocking API for testing purposes, which would allow exception tests to always pass regardless of the configuration.